[AIRFLOW-3895] GoogleCloudStorageHook create_bucket with optional request body

judoole opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

When creating a new bucket in Google Cloud Storage, GoogleCloudStorageHook.create_bucket, I would like to have the option of supplying missing parameters.

For instance, the field [lifecycle](https://cloud.google.com/storage/docs/json_api/v1/buckets#lifecycle) is not available. 

My suggestion to supply this is to add an optional dict to the parameters that gets overwritten with bucket_name, storage_class and collection.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- Updates `GoogleCloudStorageCreateBucketTest` with `resource` param
- Added two tests for `GoogleCloudStorageHook.create_bucket`

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

judoole committed:
Add a optional dict resource to create_bucket

This makes it possible to supply missing request-parameters when
creating a bucket in Google Cloud Storage

zhongjiajie commented on Dockerfile:
```diff
@@ -36,6 +36,7 @@ RUN set -x \
 
 WORKDIR $AIRFLOW_HOME
 RUN mkdir -p $AIRFLOW_HOME
+RUN airflow initdb
```
Do not think we need `airflow initdb` here. L41 had `ENTRYPOINT`
https://github.com/apache/airflow/blob/067a1e3f4aa153195674950fee72bdd23652e61d/scripts/docker/entrypoint.sh#L24-L26
which exec `airflow` command including `airflow initdb`
```suggestion
```

zhongjiajie reviewed (commented):
Do not need airflow initdb here

judoole committed:
Add param to GoogleCloudStorageCreateBucketOperator

This makes use of the new resource param in
GoogleCloudStorageCreateBucketOperator

ryanyuan commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -543,11 +549,12 @@ def create_bucket(self,
 
         service = self.get_conn()
-        bucket_resource = {
+        bucket_resource = copy.copy(resource)
```
Does it have to copy the _resource_ to the _bucket_resource_?

ryanyuan commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -543,11 +549,12 @@ def create_bucket(self,
         service = self.get_conn()
-        bucket_resource = {
+        bucket_resource = copy.copy(resource)
+        bucket_resource.update({
```
If _resource_ is None, _bucket_resource_ is None. It can't update a NoneType object.

ryanyuan commented on tests/contrib/operators/test_gcs_operator.py:
```diff
@@ -42,6 +42,12 @@ def test_execute(self, mock_hook):
+                          "condition": {"age": 7}
+                          }
+                         ]}
+            },
```
```suggestion
                      },
```

judoole committed:
Update tests and remove use of copy

GoogleCloudStorageHook.create_bucket should not use copy, but instead
just create an empty dict if resource is None.
Added a test for the create_bucket method that invokes a mock and
asserts correct invocation.

judoole committed:
Merge branch 'master' into AIRFLOW-3895-GoogleCloudStorageHook-create-bucket

judoole committed:
Fix flake8 errors

judoole committed:
Add missing mock annotation
